include/fmt/format-inl.h

```diff
@@ -683,7 +683,7 @@ template <int GRISU_VERSION> struct grisu_shortest_handler {
 };
 
 template <typename Double, FMT_ENABLE_IF_T(sizeof(Double) == sizeof(uint64_t))>
-FMT_FUNC bool grisu_format(Double value, buffer<char>& buf, int precision,
+FMT_FUNC FMT_API bool grisu_format(Double value, buffer<char>& buf, int precision,
```

Comment:
`FMT_API` shouldn't be here because it's an internal symbol.

Reply:
In DLL mode, grisu_format is refered by custom-formatter-test.cc.
```
C:\Users\User\AppData\Roaming\fmt>ninja -j 2
[3/47] Linking CXX shared library bin\fmt.dll
   Creating library fmt.lib and object fmt.exp
[12/47] Linking CXX executable bin\custom-formatter-test.exe
FAILED: bin/custom-formatter-test.exe
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=test\CMakeFiles\custom-formatter-test.dir --rc=C:\PROGRA~2\WINDOW~4\10\bin\100177~1.0\x64\rc.exe --mt=C:\PROGRA~2\WINDOW~4\10\bin\100177~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MIB055~1\2019\BUILDT~1\VC\Tools\MSVC\1420~1.275\bin\Hostx64\x64\link.exe /nologo test\CMakeFiles\custom-formatter-test.dir\custom-formatter-test.cc.obj  /out:bin\custom-formatter-test.exe /implib:test\custom-formatter-test.lib /pdb:bin\custom-formatter-test.pdb /version:0.0  /machine:x64 /INCREMENTAL:NO  /subsystem:console  test\test-main.lib test\gmock.lib fmt.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK: command "C:\PROGRA~2\MIB055~1\2019\BUILDT~1\VC\Tools\MSVC\1420~1.275\bin\Hostx64\x64\link.exe /nologo test\CMakeFiles\custom-formatter-test.dir\custom-formatter-test.cc.obj /out:bin\custom-formatter-test.exe /implib:test\custom-formatter-test.lib /pdb:bin\custom-formatter-test.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console test\test-main.lib test\gmock.lib fmt.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:bin\custom-formatter-test.exe.manifest" failed (exit code 1120) with the following output:
custom-formatter-test.cc.obj : error LNK2019: unresolved external symbol "bool __cdecl fmt::v5::internal::grisu_format<double,0>(double,class fmt::v5::internal::buffer<char> &,int,unsigned int,int &)" (??$grisu_format@N$0A@@internal@v5@fmt@@YA_NNAEAV?$buffer@D@012@HIAEAH@Z) referenced in function "private: void __cdecl fmt::v5::basic_writer<class fmt::v5::back_insert_range<class fmt::v5::internal::buffer<char> > >::write_double<double>(double,struct fmt::v5::basic_format_specs<char> const &)" (??$write_double@N@?$basic_writer@V?$back_insert_range@V?$buffer@D@internal@v5@fmt@@@v5@fmt@@@v5@fmt@@AEAAXNAEBU?$basic_format_specs@D@12@@Z)
bin\custom-formatter-test.exe : fatal error LNK1120: 1 unresolved externals
[13/47] Building CXX object test\CMakeFiles\format-test.dir\format-test.cc.obj
ninja: build stopped: subcommand failed.
```

Reply:
Ah, you are right, it should be exported. However `FMT_API` is already on the declaration, so it shouldn't be needed here: https://github.com/fmtlib/fmt/blob/ea2976e6d5a6c707d218566ef913dd3f7a93edc1/include/fmt/format.h#L1125. I wonder why MSVC is complaining.

Reply:
Anyway, if it fixes the problem, it's fine to keep `FMT_API` here but please remove `FMT_FUNC` then.

Reply:
I've also removed FMT_API from the corresponding declaration in `format.h`
```
template <typename Double, FMT_ENABLE_IF(sizeof(Double) == sizeof(uint64_t))>
FMT_API bool grisu_format(Double, buffer<char>&, int, unsigned, int&);
```

Now it becomes
```
template <typename Double, FMT_ENABLE_IF(sizeof(Double) == sizeof(uint64_t))>
bool grisu_format(Double, buffer<char>&, int, unsigned, int&);
```

Reply:
Sorry I didn't refresh this page. Now I'm looking at it.

Reply:
In `format.h`, it's been rolled back to
```
template <typename Double, FMT_ENABLE_IF(sizeof(Double) == sizeof(uint64_t))>
FMT_API bool grisu_format(Double, buffer<char>&, int, unsigned, int&);
```

In `format-inl.h`, remove FMT_FUNC and add FMT_API
```
template <typename Double, typename std::enable_if<
                               sizeof(Double) == sizeof(uint64_t), int>::type>
FMT_API bool grisu_format(Double value, buffer<char>& buf, int precision,
                           unsigned options, int& exp) {
```

Reply:
In DLL mode, `format-test.cc` refers to
```
internal::count_digits<4,struct fmt::v5::internal::uintptr::uintptr_t>(struct fmt::v5::internal::uintptr::uintptr_t)
```
The linker can not find symbol :
```
C:\Users\User\AppData\Roaming\fmt>ninja -j 2
[3/47] Linking CXX shared library bin\fmt.dll
   Creating library fmt.lib and object fmt.exp
[15/47] Linking CXX executable bin\format-test.exe
FAILED: bin/format-test.exe
cmd.exe /C "cd . && "C:\Program Files\CMake\bin\cmake.exe" -E vs_link_exe --intdir=test\CMakeFiles\format-test.dir --rc=C:\PROGRA~2\WINDOW~4\10\bin\100177~1.0\x64\rc.exe --mt=C:\PROGRA~2\WINDOW~4\10\bin\100177~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MIB055~1\2019\BUILDT~1\VC\Tools\MSVC\1420~1.275\bin\Hostx64\x64\link.exe /nologo test\CMakeFiles\format-test.dir\format-test.cc.obj  /out:bin\format-test.exe /implib:test\format-test.lib /pdb:bin\format-test.pdb /version:0.0  /machine:x64 /INCREMENTAL:NO  /subsystem:console  test\test-main.lib test\gmock.lib fmt.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
LINK: command "C:\PROGRA~2\MIB055~1\2019\BUILDT~1\VC\Tools\MSVC\1420~1.275\bin\Hostx64\x64\link.exe /nologo test\CMakeFiles\format-test.dir\format-test.cc.obj /out:bin\format-test.exe /implib:test\format-test.lib /pdb:bin\format-test.pdb /version:0.0 /machine:x64 /INCREMENTAL:NO /subsystem:console test\test-main.lib test\gmock.lib fmt.lib kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTFILE:bin\format-test.exe.manifest" failed (exit code 1120) with the following output:
format-test.cc.obj : error LNK2019: unresolved external symbol "int __cdecl fmt::v5::internal::count_digits<4,struct fmt::v5::internal::uintptr::uintptr_t>(struct fmt::v5::internal::uintptr::uintptr_t)" (??$count_digits@$03Uuintptr_t@uintptr@internal@v5@fmt@@@internal@v5@fmt@@YAHUuintptr_t@uintptr@012@@Z) referenced in function "public: void __cdecl fmt::v5::basic_writer<class fmt::v5::back_insert_range<class fmt::v5::internal::buffer<char> > >::write_pointer<struct fmt::v5::internal::uintptr::uintptr_t>(struct fmt::v5::internal::uintptr::uintptr_t,struct fmt::v5::align_spec const *)" (??$write_pointer@Uuintptr_t@uintptr@internal@v5@fmt@@@?$basic_writer@V?$back_insert_range@V?$buffer@D@internal@v5@fmt@@@v5@fmt@@@v5@fmt@@QEAAXUuintptr_t@uintptr@internal@12@PEBUalign_spec@12@@Z)
bin\format-test.exe : fatal error LNK1120: 1 unresolved externals
[16/47] Building CXX object test\CMakeFiles\assert-test.dir\assert-test.cc.obj
ninja: build stopped: subcommand failed.
```

In format-test.cc , if I comment out
```
TEST(WriterTest, WriteUIntPtr) {
  memory_buffer buf;
  fmt::writer writer(buf);
  writer.write_pointer(fmt::internal::bit_cast<fmt::internal::uintptr_t>(
                           reinterpret_cast<void*>(0xface)),
                       FMT_NULL);
  EXPECT_EQ("0xface", to_string(buf));
}
```
All test are passed.

Reply:
Oh, it's already been fixed. I didn't see. Thank you.